test/test_bundle_importer.js

```diff
@@ -191,26 +213,70 @@ describe('Bundle Importer', function() {
         addMachines: true,
         deploy: true,
         addUnit: true,
-        addRelation: true
+        addRelation: true,
+        setAnnotations: true
       });
     });
-
-    it('does not fail if unknown record type', function() {
+    it('properly sorts a recordSet', function() {
+      var data = utils.loadFixture(
+          'data/wordpress-bundle-recordset.json', true);
+      var sorted = bundleImporter._sortDryRunRecords(data);
+      var order = [
+        'addCharm-0', 'addService-1', 'setAnnotations-2',
+        'addCharm-3', 'addService-4', 'setAnnotations-5',
+        'addCharm-6', 'addService-7', 'setAnnotations-8',
+        'addMachines-9', 'addMachines-10',
+        'addRelation-11', 'addRelation-12',
+        'addUnit-13', 'addMachines-16',
+        'addUnit-14', 'addMachines-17',
+        'addUnit-15'
+      ];
+      sorted.forEach(function(record, index) {
+        assert.equal(record.id, order[index]);
+      });
+    });
+    it('stops but does not fail if unknown record type', function() {
       var sortedRecords = [
         { method: 'badMethod' }
       ];
+      var execute = utils.makeStubMethod(bundleImporter, '_executeRecord');
+      var notification = utils.makeStubMethod(
+          bundleImporter.db.notifications, 'add');
+      this._cleanups.concat([execute.reset, notification.reset]);
       bundleImporter._executeDryRun(sortedRecords);
+      execute.passThroughToOriginalMethod();
+      // the executor should only be called once at which time it'll throw a
+      // notification instead of continuing on.
+      assert.equal(execute.callCount(), 1, 'execute not called');
+      assert.equal(notification.callCount(), 1, 'notification not added');
     });
+  });
 
-    describe('_execute_addCharm', function() {
-      it('loads a charm');
-      it('shows notification on failure');
+  describe('Changeset execution', function() {
+    it('Sets up the correct environment (Integration)', function() {
+      var data = utils.loadFixture(
+          'data/wordpress-bundle-recordset.json', true);
+      bundleImporter.importBundleDryRun(data);
+      assert.equal(db.services.size(), 3);
```

Comment:
This test assumes that the db, env, and fakebackend are all working correctly. Each executor calls out to a number of env, db, fakebackend methods which would have made the tests very tedious as we'd have to re-implement the api and return structures.